### ISSUE TITLE
AddStdCatalog() prefers versioned wxstd catalogs in a default path to unversioned ones in a user-specified path

### ISSUE THREAD

**Issue by taler21** [wxWidgets/wxWidgets#22501]
The documentation of `wxTranslations::AddCatalog(...)` states:
> By default, it is searched for in standard places (see wxFileTranslationsLoader), but you may also prepend additional directories to the search path with wxFileTranslationsLoader::AddCatalogLookupPathPrefix().

Even if this is not explicitly stated for the standard wxWidgets catalogs, this also applied to `wxTranslations::AddStdCatalog()` in the past.

In wxWidgets-3.1.5, commit ded4da5ce50 introduced wxWidgets message catalogs with version suffix.
And since then `wxTranslations::AddStdCatalog()` prefers versioned wxstd catalogs in a default path to unversioned ones in a user-specified path.

The current search logic is as follows:
1. `wxstd-3.1.mo` in additional directories prepended with `wxFileTranslationsLoader::AddCatalogLookupPathPrefix()`
2. `wxstd-3.1.mo` in standard directories
3. `wxstd.mo` in additional directories prepended with `wxFileTranslationsLoader::AddCatalogLookupPathPrefix()`
4. `wxstd.mo` in standard directories

But the correct search logic (from my point of view) would be the following:
1. `wxstd-3.1.mo` in additional directories prepended with `wxFileTranslationsLoader::AddCatalogLookupPathPrefix()`
2. `wxstd.mo` in additional directories prepended with `wxFileTranslationsLoader::AddCatalogLookupPathPrefix()`
3. `wxstd-3.1.mo` in standard directories
4. `wxstd.mo` in standard directories

Of course, `AddCatalog("wxstd")` could be used instead of `AddStdCatalog()` to ensure that unversioned wxstd catalogs that come with the application and are located in an application-specific path are used.
But I think it would be better if `AddStdCatalog()` could still be used.
Especially since the documentation of `wxTranslations::AddStdCatalog()` also mentions possible port-specific catalogs.
> Add standard wxWidgets catalogs ("wxstd" and possible port-specific catalogs).

**Platform and version information**
 - wxWidgets version you use: 3.1.7
 - wxWidgets port you use: wxGTK

**Comment by vadz** [comment]
Just to be clear, the intention here is to override the standard translations with your own, right? If so, I wonder if this wouldn't be achieved more clearly and reliably by using (the not yet existing) `SetCatalogLookupPathPrefix()` rather than `Add()` and not looking in the standard locations at all?

**Comment by taler21** [comment]
Yes, I have to make sure that compatible catalogs are always used.
And from my point of view this can only be guaranteed if suitable standard wxWidgets catalogs are supplied with the application for each platform, including Linux.
Especially since release versions of our applications also use development versions of wxWidgets (i.e. versions 3.1.x).
And for incompatible changes such as 3396a400152ba84f32e35dfa2f87af21e97d9657, between 3.1.6 and 3.1.7, it is essential that the application uses the catalog appropriate to the version used, and not any installed on the system with the 3.1 version suffix.

So in principle a `SetCatalogLookupPathPrefix()` would achieve that, yes.
But how would it handle the `possible port-specific catalogs` mentioned in the documentation (if there will ever be such port-specific catalogs)?

Maybe it would be the best to change `AddStdCatalog()`'s search logic to generally prefer message catalogs in user-specific paths, as it was the case up to and including wxWidgets-3.1.4.
From my perspective, its current search logic is neither what is expected nor makes much sense.
Why should `AddStdCatalog()` prefer versioned wxstd catalogs in a default path to unversioned catalogs in an application-specific path?
In my opinion, wxstd catalogs in an application-specific path should always be preferred, whether versioned or not.

And changing `AddStdCatalog()`'s current search logic would also avoid breaking the behavior in that respect from wxWidgets-3.0 to wxWidgets-3.2.
But maybe I am the only one using `AddCatalogLookupPathPrefix()` and `AddStdCatalog()` this way, I don't know.

If you think `SetCatalogLookupPathPrefix()` is the way to go, I will used it (as soon as available).

**Comment by vadz** [comment]
I think that if you're using your own catalogs, you'd have to provide any port-specific catalogs too.

But thinking more about it, I wonder why you don't just call your catalogs `wxstd-3.1.mo`? It would seem that this should allow them to work right now, without any changes.

**Comment by vadz** [comment]
> It seems that platform-specific translations have been droped with commit [26eb738](https://github.com/wxWidgets/wxWidgets/commit/26eb738562d978901dfef500af5707184290e62a). Do you think there will ever be platform-specific translations again? 

In principle, having platform-dependent translations makes sense, but in practice we've never really used it much and it just wasn't worth the extra trouble. Also, I don't know of any other applications using this, and it doesn't look like a good idea to do things differently from everybody else without some really good reason.

> If not, I could probably just call `AddCatalog("wxstd")` instead of `AddStdCatalog()` to get the behavior I want. Or do you see any potential problems doing it that way?

No, `AddCatalog()` should work fine.
 
> I am still convinced that the current behavior of `AddStdCatalog()` is neither optimal nor correct.

I probably agree but I'd rather avoid changing this now and risk breaking something else...
